cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java

```diff
@@ -423,12 +424,47 @@ public RunRecordDetail getRun(ProgramRunId id) {
   }
 
   @Override
-  public void addApplication(ApplicationId id, ApplicationSpecification spec) {
+  public void addApplication(ApplicationId id, ApplicationSpecification spec, @Nullable String author,
+                             Long created, @Nullable String changeSummary, @Nullable String parentVersion) {
+    // Generate the creation time for the version of the app.
     TransactionRunners.run(transactionRunner, context -> {
-      getAppMetadataStore(context).writeApplication(id.getNamespace(), id.getApplication(), id.getVersion(), spec);
+      ApplicationMeta latest = getLatest(id.getNamespace(), id.getApplication());
+      if (!deployAppAllowed(parentVersion, latest)) {
+        throw new BadRequestException("Can't deploy the application because the parent version is not the " +
+                                        "latest.");
+      }
+      // When the app does not exist -it is not an edit).
+      if (latest != null) {
+        getAppMetadataStore(context).updateLatestApplication(id.getNamespace(), id.getApplication(),
```

Comment:
save getAppMetadataStore(context) in a variable so it can be called just once instead of both here and below.

Reply:
we are moving this to AppMetadataStore, so this no longer applies